okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -185,81 +212,125 @@ public void onResponse(Call call, Response response) throws IOException {
                     // so error handlers can read the body without breaking subsequent handlers
                     errorResponseSupplier = () -> buildFrom(response, body);
                 } catch (IOException e) {
-                    onFailure(call, e);
+                    completion.onError(call, e);
                     return;
                 }
 
                 // Handle to handle QoS situations: retry, failover, etc.
                 Optional<QosException> qosError = qosHandler.handle(errorResponseSupplier.get());
                 if (qosError.isPresent()) {
-                    qosError.get().accept(createQosVisitor(callback, call));
+                    qosError.get().accept(createQosVisitor(call, completion));
                     return;
                 }
 
                 // Handle responses that correspond to RemoteExceptions / SerializableErrors
                 Optional<RemoteException> httpError = remoteExceptionHandler.handle(errorResponseSupplier.get());
                 if (httpError.isPresent()) {
-                    callback.onFailure(call, new IoRemoteException(httpError.get()));
+                    completion.onError(call, new IoRemoteException(httpError.get()));
                     return;
                 }
 
                 // Catch-all: handle all other responses
                 Optional<IOException> ioException = ioExceptionHandler.handle(errorResponseSupplier.get());
                 if (ioException.isPresent()) {
-                    callback.onFailure(call, ioException.get());
+                    completion.onError(call, ioException.get());
                     return;
                 }
 
-                callback.onFailure(call, new IOException("Failed to handle request, this is an http-remoting bug."));
+                completion.onError(call, new IOException("Failed to handle request, this is an http-remoting bug."));
             }
         });
+
     }
 
-    @SuppressWarnings("FutureReturnValueIgnored")
-    private void scheduleExecution(Runnable execution, Duration backoff) {
-        // TODO(rfink): Investigate whether ignoring the ScheduledFuture is safe, #629.
-        schedulingExecutor.schedule(
-                () -> executionExecutor.submit(execution),
-                backoff.toMillis(),
-                TimeUnit.MILLISECONDS);
+    private static final class RequestCompletion {
+        private final ScheduledExecutorService schedulingExecutor;
+        private final ExecutorService executionExecutor;
+        private final Callback callback;
+        private final Listener listener;
+
+        private RequestCompletion(
+                ScheduledExecutorService schedulingExecutor, ExecutorService executionExecutor,
```

Comment:
could make this class not static and just reference the executors directly

Reply:
done